packages/expo-ota/android/src/main/java/expo/modules/ota/OtaUpdater.kt

```diff
@@ -0,0 +1,188 @@
+    }
+
+    fun downloadManifest(success: (JSONObject) -> Unit, error: (Exception) -> Unit) {
+        config.manifestHttpClient.newCall(createManifestRequest(config)).enqueue(object : Callback {
+            override fun onFailure(call: Call, e: IOException) {
+                error(IllegalStateException("Manifest fetching failed: ", e))
+            }
+
+            override fun onResponse(call: Call, response: Response) {
+                if (response.isSuccessful) {
+                    if (response.body() != null) {
+                        verifyManifest(response, success, error)
+                    } else {
+                        error(IllegalStateException("Response body is null: ", response.body()))
+                    }
+                } else {
+                    error(IllegalStateException("Response not successful. Code: " + response.code() + ", body: " + response.body()?.toString()))
+                }
+            }
+        })
+    }
+
+    fun verifyManifest(response: Response, success: (JSONObject) -> Unit, error: (Exception) -> Unit) {
+        config.manifestResponseValidator.validate(response, {
+            success(JSONObject(it))
+        }, error)
+    }
+
+    fun saveDownloadedManifestAndBundlePath(manifest: JSONObject, path: String) {
+        val previousBundle = persistence.downloadedBundlePath
+        if (previousBundle != null) {
+            removeFile(previousBundle) // TODO: Move to persistence!
+        }
+        persistence.downloadedManifest = manifest
+        persistence.downloadedBundlePath = path
+    }
+
+    fun prepareToReload() {
+        persistence.enqueuedReorderAtNextBoot = true
+        persistence.synchronize()
+    }
+
+    fun markDownloadedCurrentAndCurrentOutdated() {
+        val outdated = persistence.outdatedBundlePath
+        if (outdated != null) {
+            removeFile(outdated)
+        }
+        persistence.markDownloadedCurrentAndCurrentOutdated()
+    }
+
+    fun removeOutdatedBundle() {
+        val outdatedBundlePath = persistence.outdatedBundlePath
+        persistence.outdatedBundlePath = null
+        if (outdatedBundlePath != null) {
+            removeFile(outdatedBundlePath)
+        }
+    }
+
+    fun cleanUnusedFiles() {
+        val bundlesDir = bundleDir()
+        if (bundlesDir.exists() && bundlesDir.isDirectory) {
+            bundlesDir.listFiles { directory, filename -> !validFilesSet.contains(File(directory, filename).path) }
+                    .forEach { removeFile(it.path) }
+        }
+    }
+
+    private fun checkAssetsBundleAndManifest() {
+        if (shouldCopyAssetsManifestAndBundle()) {
+            val embeddedManifest: JSONObject = embeddedManifestAndBundle.readManifest()
+            saveResponseToFile(bundleDir(), bundleFilename(embeddedManifest)) (embeddedManifestAndBundle.readBundle(), {
+                saveDownloadedManifestAndBundlePath(embeddedManifest, it)
+                markDownloadedCurrentAndCurrentOutdated()
```

Comment:
is it possible that we want to set `persistence.enqueuedReorderAtNextBoot = false` here? Seems like it might be safe since if we get to this point we know which bundle we are going to run. Will defer to you on this though.